src/Build/Graph/ProjectGraph.cs

```diff
@@ -175,31 +222,38 @@ public ProjectGraph(string entryProjectFile)
         /// </summary>
-        private void LoadGraph(string entryProjectFile, ProjectCollection projectCollection, Dictionary<string, string> globalProperties, string toolsVersion)
+        private void LoadGraph(string entryProjectFile, ProjectCollection projectCollection, PropertyDictionary<ProjectPropertyInstance> globalProperties, string toolsVersion)
         {
-            var projectsToEvaluate = new Queue<KeyValuePair<string, ProjectGraphNode>>();
+            var projectsToEvaluate = new Queue<KeyValuePair<ConfigurationMetadata, ProjectGraphNode>>();
             // entry project node has no parent
-            projectsToEvaluate.Enqueue(new KeyValuePair<string, ProjectGraphNode>(entryProjectFile, null));
+            projectsToEvaluate.Enqueue(new KeyValuePair<ConfigurationMetadata, ProjectGraphNode>(new ConfigurationMetadata(entryProjectFile, toolsVersion, globalProperties), null));
             while (projectsToEvaluate.Count != 0)
             {
-                KeyValuePair<string, ProjectGraphNode> projectToEvaluateAndParentNode = projectsToEvaluate.Dequeue();
-                string projectToEvaluate = projectToEvaluateAndParentNode.Key;
+                KeyValuePair<ConfigurationMetadata, ProjectGraphNode> projectToEvaluateAndParentNode = projectsToEvaluate.Dequeue();
+                ConfigurationMetadata projectToEvaluate = projectToEvaluateAndParentNode.Key;
+                ProjectGraphNode parentNode = projectToEvaluateAndParentNode.Value;
+
                 if (!_allParsedProjects.TryGetValue(projectToEvaluate, out ProjectGraphNode parsedProject))
                 {
-                    parsedProject = CreateNewNode(projectToEvaluate, projectCollection, globalProperties, toolsVersion);
+                    parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
                     IEnumerable<ProjectItemInstance> projectReferenceItems = parsedProject.Project.GetItems(ProjectReferenceItemName);
                     foreach (var projectReferenceToParse in projectReferenceItems)
                     {
-                        string projectReferencePath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
-                        projectsToEvaluate.Enqueue(new KeyValuePair<string, ProjectGraphNode>(projectReferencePath, parsedProject));
+                        string projectReferenceFullPath = projectReferenceToParse.GetMetadataValue(FullPathMetadataName);
+                        string projectReferenceToolsVersion = projectReferenceToParse.GetMetadataValue(ToolsVersionMetadataName);
+                        if (string.IsNullOrEmpty(projectReferenceToolsVersion))
+                        {
+                            // TODO: This should actually be the project reference's tools version, but we haven't actually laoded it yet
+                            projectReferenceToolsVersion = toolsVersion;
```

Comment:
I'd lean more toward "throw if they tried to set ToolsVersion".

Reply:
Sure. In that case we can just hardcode something in `ConfigurationMetadata` (or just always use the ProjectCollection's default I guess) and just assume that at execution time the one specified in the project is the one that's actually used caching-wise.